File: Algo/Leetcode/098ValidBST.cpp
// use a pointer
class Solution {
public:
    bool isBST = true;
    TreeNode* pre = NULL;
    bool isValidBST(TreeNode* root) {
        if(!root) return isBST;
        dfs(root);
        return isBST;
    }

    void dfs(TreeNode* root){
        if(!root) return;

        dfs(root->left);
        if(!pre) pre = root;
        else {
            if(pre->val >= root->val)
                isBST = false;
        }
        pre = root;
        dfs(root->right);
    }
};

// use a single value. But this is not correct if some value is INT_MIN in the tree.
class Solution {
public:
    bool isBST = true;
    int last_val = INT_MIN;
    bool isValidBST(TreeNode* root) {
        if(!root) return isBST;
        dfs(root);
        return isBST;
    }

    void dfs(TreeNode* root){
        if(!root) return;

        dfs(root->left);
        if(last_val >= root->val) isBST = false;
        last_val = root->val;
        dfs(root->right);
    }
};

File: Algo/Leetcode/098ValidBST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):

    def __init__(self):
        self.last_min = None
        self.bst = True

    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None:
            return True

        self.dfs(root)

        return self.bst

    def dfs(self, root):
        if root is None:
            return

        self.dfs(root.left)
        if self.last_min is not None and root.val <= self.last_min:
            self.bst = False
        self.last_min = root.val
        self.dfs(root.right)